packages/devtools/lib/src/timeline/cpu_profile_protocol.dart

```diff
@@ -98,3 +99,107 @@ class CpuProfileProtocol {
+
+    // Merge samples when possible starting at the root (the leaf node of the
+    // original CPU sample).
+    mergeRoots(bottomUpRoots);
+
+    return bottomUpRoots;
+  }
+
+  /// Returns the roots for a bottom up representation of a CpuStackFrame node.
+  ///
+  /// Each root is a leaf from the original CpuStackFrame tree, and its children
+  /// will be the reverse call stack of the original sample. The stack frames
+  /// returned will not be merged to combine common roots, and the sample counts
+  /// will not reflect the bottom up sample counts. These steps will occur later
+  /// in the bottom-up conversion process.
+  @visibleForTesting
+  List<CpuStackFrame> getRoots(
+    CpuStackFrame node,
+    CpuStackFrame currentBottomUpRoot,
+    List<CpuStackFrame> bottomUpRoots,
+  ) {
+    final copy = node.shallowCopy()
+      ..exclusiveSampleCount = node.exclusiveSampleCount;
+
+    if (currentBottomUpRoot != null) {
+      copy.addChild(currentBottomUpRoot.deepCopy());
+    }
+
+    // [copy] is the new root of the bottom up call stack.
+    currentBottomUpRoot = copy;
+
+    if (node.exclusiveSampleCount > 0) {
+      // This node is a leaf node, meaning it is a bottom up root.
+      bottomUpRoots.add(currentBottomUpRoot);
+    }
+    for (CpuStackFrame child in node.children) {
+      getRoots(child, currentBottomUpRoot, bottomUpRoots);
+    }
+    return bottomUpRoots;
+  }
+
+  /// Sets sample counts of [stackFrame] and all children to
+  /// [exclusiveSampleCount].
+  ///
+  /// This is necessary for the transformation of a [CpuStackFrame] to its
+  /// bottom-up representation. This is an intermediate step between
+  /// [getRoots] and [mergeRoots].
+  @visibleForTesting
+  void cascadeSampleCounts(CpuStackFrame stackFrame) {
+    stackFrame.inclusiveSampleCount = stackFrame.exclusiveSampleCount;
+    for (CpuStackFrame child in stackFrame.children) {
+      child.exclusiveSampleCount = stackFrame.exclusiveSampleCount;
+      cascadeSampleCounts(child);
+    }
+  }
+
+  /// Merges bottom up roots that share a common call stack (starting at the
+  /// root).
+  ///
+  /// Ex. C               C                     C
+  ///      -> B             -> B        -->      -> B
+  ///          -> A             -> D                 -> A
+  ///                                                -> D
+  ///
+  /// At the time this method is called, we assume we have a list of roots with
+  /// accurate bottom up sample counts.
+  @visibleForTesting
+  void mergeRoots(List<CpuStackFrame> roots) {
+    // Loop through a copy of [roots] so that we can remove nodes from [roots]
+    // once we have merged them.
+    final List<CpuStackFrame> rootsCopy = List.from(roots);
+    for (CpuStackFrame root in rootsCopy) {
+      if (!roots.contains(root)) {
+        // We have already merged [root] and removed it from [roots]. Do not
+        // attempt to merge again.
+        continue;
+      }
+
+      final duplicates = roots
```

Comment:
nit: maybe these are matching instead of duplicate?

Reply:
renamed to `matchingRoots`